packages/merge-styles/src/mergeStyleSets.ts

```diff
@@ -1,43 +1,111 @@
 import { styleToRegistration, applyRegistration } from './styleToClassName';
+import { IStyleFunction } from './IStyleFunction';
+import { IStyleSet } from './IStyleSet';
+
+export type IClassNameOrStyleFunction<TStyleProps = any, TStyles = any> = string | IStyleFunction<TStyleProps, TStyles>;
+
+/**
+ * Takes in one or more style set objects, each consisting of a set of areas,
+ * each which will produce a class name. Using this is analogous to calling
+ * `mergeStyles` for each property in the object, but ensures we maintain the
+ * set ordering when multiple style sets are merged.
+ *
+ * @param styleSet1 The first style set to be merged.
+ */
+export function mergeStyleSets<T>(styleSet: IStyleSet<T>): { [P in keyof T]?: IClassNameOrStyleFunction };
 
 /**
- * Allows you to pass in 1 or more sets of areas which will return a merged
- * set of classes.
+ * Takes in one or more style set objects, each consisting of a set of areas,
+ * each which will produce a class name. Using this is analogous to calling
+ * `mergeStyles` for each property in the object, but ensures we maintain the
+ * set ordering when multiple style sets are merged.
  *
- * @public
+ * @param styleSet1 The first style set to be merged.
+ * @param styleSet2 The second style set to be merged.
  */
-export function mergeStyleSets<K extends string>(
-  ...cssSets: ({ [P in K]?: IStyle } | null | undefined)[]
-): { [P in K]: string } {
+export function mergeStyleSets<T, U>(
+  styleSet1: IStyleSet<T>,
+  styleSet2: IStyleSet<U>
+): { [P in keyof (T & U)]?: IClassNameOrStyleFunction };
+
+/**
+ * Takes in one or more style set objects, each consisting of a set of areas,
+ * each which will produce a class name. Using this is analogous to calling
+ * `mergeStyles` for each property in the object, but ensures we maintain the
+ * set ordering when multiple style sets are merged.
+ *
+ * @param styleSet1 The first style set to be merged.
+ * @param styleSet2 The second style set to be merged.
+ * @param styleSet3 The third style set to be merged.
+ */
+export function mergeStyleSets<T, U, V>(
+  styleSet1: IStyleSet<T>,
+  styleSet2: IStyleSet<U>,
+  styleSet3: IStyleSet<V>
+): { [P in keyof (T & U & V)]?: IClassNameOrStyleFunction };
+
+/**
+ * Takes in one or more style set objects, each consisting of a set of areas,
+ * each which will produce a class name. Using this is analogous to calling
+ * `mergeStyles` for each property in the object, but ensures we maintain the
+ * set ordering when multiple style sets are merged.
+ *
+ * @param styleSet1 The first style set to be merged.
+ * @param styleSet2 The second style set to be merged.
+ * @param styleSet3 The third style set to be merged.
+ * @param styleSet4 The fourth style set to be merged.
+ */
+export function mergeStyleSets<T, U, V, W>(
+  styleSet1: IStyleSet<T>,
+  styleSet2: IStyleSet<U>,
+  styleSet3: IStyleSet<V>,
+  styleSet4: IStyleSet<W>
+): { [P in keyof (T & U & V & W)]?: IClassNameOrStyleFunction };
```

Comment:
I'm assuming the breakout is for TS support? Would something like https://github.com/Microsoft/TypeScript/issues/3870 eventually let us collapse this?

Reply:
Maybe, depends on what they land on at the end of the day. The language otherwise isn't expressive enough today and some of the uses of any are inevitable.